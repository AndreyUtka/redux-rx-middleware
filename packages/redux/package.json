{
  "name": "@rx-dispatcher/redux",
  "version": "0.0.1",
  "scripts": {
    "test": "NODE_ENV=test jest --config ../../scripts/jest.config.js",
    "test:watch": "npm run test -- --watch",
    "test:coverage": "npm run test -- --coverage"
  },
  "dependencies": {
    "flux-standard-action": "^2.0.1",
    "redux": "^3.7.2",
    "rxjs": "^5.5.6"
  }
}
