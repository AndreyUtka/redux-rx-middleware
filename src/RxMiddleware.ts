import { FSA } from "flux-standard-action";
import { Action, Dispatch } from "redux";
import { Observable } from "rxjs";

type ObservableAction<TObservable> = FSA<Observable<TObservable>, IMeta>;
type TypedAction = Action | FSA<any> | ObservableAction<any>;
type RxMiddleware = <S>() => (next: Dispatch<S>) => <A extends TypedAction>(action: A) => A;

interface IMeta {
    sequence: Sequence;
}

export enum Sequence {
    Next = "next",
    Error = "error",
    Complete = "complete",
}

export const rxMiddleware: RxMiddleware = () => (next) => <A extends TypedAction>(action: A): A => {
    if ((action as FSA<any>).payload instanceof Observable) {
        (action as ObservableAction<any>).payload.subscribe({
            next: (data) => next({ type: action.type, payload: data, meta: { sequence: Sequence.Next } }),
            error: (err) => next({ type: action.type, payload: err, error: true, meta: { sequence: Sequence.Error } }),
            complete: () => next({ type: action.type, meta: { sequence: Sequence.Complete } }),
        });
        return action;
    }

    return next(action);
};
