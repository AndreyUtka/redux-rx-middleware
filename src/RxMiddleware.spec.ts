import { Dispatch } from "redux";
import { Observable } from "rxjs";
import { rxMiddleware } from "./RxMiddleware";

describe("RxMiddleware", () => {
    let mockDispatch: jest.Mock;
    let dispatch: Dispatch<any>;

    beforeEach(() => {
        mockDispatch = jest.fn();
        dispatch = rxMiddleware()(mockDispatch);
    });

    it("handles Flux standard actions", async () => {
        await dispatch({
            type: "ACTION_TYPE",
            payload: Observable.from([1, 2, 3]),
        });

        expect(mockDispatch.mock.calls).toMatchSnapshot();
    });

    it("ignores non-observables", async () => {
        dispatch({ type: "ACTION_TYPE" });
        expect(mockDispatch.mock.calls).toMatchSnapshot();
    });
});
