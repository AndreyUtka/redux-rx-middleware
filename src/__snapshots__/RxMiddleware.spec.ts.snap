// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RxMiddleware handles Flux standard actions 1`] = `
Array [
  Array [
    Object {
      "meta": Object {
        "sequence": "next",
      },
      "payload": 1,
      "type": "ACTION_TYPE",
    },
  ],
  Array [
    Object {
      "meta": Object {
        "sequence": "next",
      },
      "payload": 2,
      "type": "ACTION_TYPE",
    },
  ],
  Array [
    Object {
      "meta": Object {
        "sequence": "next",
      },
      "payload": 3,
      "type": "ACTION_TYPE",
    },
  ],
  Array [
    Object {
      "meta": Object {
        "sequence": "complete",
      },
      "type": "ACTION_TYPE",
    },
  ],
]
`;

exports[`RxMiddleware ignores non-observables 1`] = `
Array [
  Array [
    Object {
      "type": "ACTION_TYPE",
    },
  ],
]
`;
